website/testCSSOrder.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const path = require('path');
+const fs = require('fs');
+
+/*
+This verifies CSS ordering on the Docusaurus site itself,
+
+There are multiple ways to provide some CSS to Docusaurus
+and Docusaurus should guarantee a consistent CSS ordering over time
+
+See also
+- https://github.com/facebook/docusaurus/issues/3678
+- https://github.com/facebook/docusaurus/pull/5987
+
+TODO we should probably add a real e2e test in core instead of using our own website?
+Current solution looks good-enough for now
+
+ */
+
+// TODO temporary, the current order is bad and we should change that
+const EXPECTED_MARKER_CLASSES = [
+  // Some class from Infima, in the order Infima declares them
+  // this ensures we don't mess-up and re-order css from other libs
+  '.markdown>h2',
+  '.button--outline.button--active',
+  ':root', // TODO should be first
+  '.DocSearch-Hit-content-wrapper', // TODO should not be there?
+  '.navbar__title',
+  '.test-marker-site-custom-css-shared-rule', // TODO should not be there!
+  '.col[class*=col--]', // TODO should be after paddings
+  '.padding-vert--xl',
+  '.footer__link-item', // TODO should be last
+  '.pagination__item',
+  '.pills__item',
+  '.tabs__item',
+
+  // Test markers
+  '.test-marker-site-custom-css-unique-rule',
+  '.test-marker-site-client-module',
+  '.test-marker-theme-layout',
+  '.test-marker-site-index-page',
+  '.DocSearch-Modal',
+];
```

Comment:
@lex111 @Josh-Cena as we can see the output CSS is currently in a weird order.

Due to CSS optimizations, Infima classes are re-ordered. Site custom CSS classes are "merged" with Infima classes sharing similar rules and put at the very beginning. We end up with CSS classes from different places (site, client modules, Infima, theme...) being interleaved in an unpredictable way.

I'm not sure if this is the correct thing to do in all cases, and I'd rather disable such optimizations, as it's better to do something that works than save a few Kb. 

Are these optimizations always guaranteed 100% safe?

---

The main problem remains that site custom CSS should rather be last

Reply:
Ah, if it's because of optimization it makes perfect sense, and that's the entire purpose of optimization: combine duplicated classes where it can. If you have two class names that are identical in every way, it will be safe to combine them. I wouldn't worry about that since there's nothing to override in this case either.

Reply:
We already have a bloated bundle size that I've been looking for ways to minimize (e.g. #3655 could be a way out). Optimizations are meant to be _optimizations_: there aren't behavior differences but you can't expect everything to be the same, in the same way you can't rely on `someFunction.name` since it can be minified